helix-core/src/main/java/org/apache/helix/controller/stages/CustomizedStateOutput.java

```diff
@@ -24,32 +24,63 @@
 import java.util.Map;
 import java.util.Set;
 
+import org.apache.helix.model.CustomizedState;
 import org.apache.helix.model.Partition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 public class CustomizedStateOutput {
+  private static Logger LOG = LoggerFactory.getLogger(CustomizedStateOutput.class);
+
   // stateType -> (resourceName -> (Partition -> (instanceName -> customizedState)))
   private final Map<String, Map<String, Map<Partition, Map<String, String>>>> _customizedStateMap;
+  // stateType -> (resourceName -> (Partition -> (instanceName -> startTime)))
+  private final Map<String, Map<String, Map<Partition, Map<String, String>>>> _startTimeMap;
 
   public CustomizedStateOutput() {
     _customizedStateMap = new HashMap<>();
+    _startTimeMap = new HashMap<>();
   }
 
   public void setCustomizedState(String stateType, String resourceName, Partition partition,
       String instanceName, String state) {
-    if (!_customizedStateMap.containsKey(stateType)) {
-      _customizedStateMap
-          .put(stateType, new HashMap<String, Map<Partition, Map<String, String>>>());
+    setCustomizedStateProperty(CustomizedState.CustomizedStateProperty.CURRENT_STATE, stateType,
+        resourceName, partition, instanceName, state);
+  }
+
+  public void setStartTime(String stateType, String resourceName, Partition partition,
+      String instanceName, String state) {
+    setCustomizedStateProperty(CustomizedState.CustomizedStateProperty.START_TIME, stateType,
+        resourceName, partition, instanceName, state);
+  }
+
+  private void setCustomizedStateProperty(CustomizedState.CustomizedStateProperty propertyName,
+      String stateType, String resourceName, Partition partition, String instanceName,
+      String state) {
+    Map<String, Map<String, Map<Partition, Map<String, String>>>> mapToUpdate;
+    switch (propertyName) {
+      case CURRENT_STATE:
+        mapToUpdate = _customizedStateMap;
+        break;
+      case START_TIME:
+        mapToUpdate = _startTimeMap;
+        break;
+      default:
+        LOG.error(
```

Comment:
Often, the only use for the ERROR level within a certain application is when a valuable business use case cannot be completed due to technical issues or a bug. 
I'd prefer not to use this logging level too generously because that would add too much noise to the logs and reduce the significance of a single ERROR event.

The WARN level should be used when something bad happened, but the application still has the chance to heal itself or the issue can wait a day or two to be fixed.

Reply:
Good to know. Thanks! I changed all the ERROR level to warning since it's not significant issue